design/thin-edge-core.md

```diff
@@ -0,0 +1,197 @@
+│                │      #      │                 │     │                 │
+└───────┬─▲──────┘      #      └──────┬─▲────────┘     └──────▲─┬────────┘
+        │ │             #             │ │                     │ │
+        │ │  SmartRest  #             │ │ JSON,CSV,SmartRest  │ │ JSON
+        │ │             #             │ │                     │ │
+┌───────▼─┴─────────────#─────────────▼─┴─────────────────────┴─▼────────────────┐
+│  MQTT                 #                                                        │
+│                       #                                                        │
+└───────────────────────#─────────────▲───────────────────────▲──────────────────┘
+                        #             │                       │
+                        #             │ CSV                   │  JSON
+                        #             │                       │
+                        #      ┌──────┴──────────┐      ┌─────┴──────────┐
+                        #      │                 │      │                │
+                        #      │ Collectd        │      │ Third-party    │
+                        #      │                 │      │                │
+                        #      │                 │      │                │
+                        #      │                 │      │                │
+                        #      │                 │      │                │
+                        #      │                 │      │                │
+                        #      └─────────────────┘      └────────────────┘
+                        #
+                        #
+```
+
+If we zoom into a built-in thin-edge executable, then we have a different kind of components, Rust actors,
+that exchange typed messages over in-memory channels.
+
+For instance, the generic mapper provides support for Cumulocity, Azure, collectd
+and telemetry data (measurements, events, alarms) collected over MQTT using the thin-edge JSON format.
+
+```
+┌──────────────────────────────────────────────────────────┐
+│                                                          │
+│ Generic Mapper                                           │
+│                                                          │
+│                                                          │
+│  ┌─────────────────┐                                     │
+│  │ c8y plugin      ├────► operations ──────┐             │
+│  │                 │                       │             │
+│  │                 │                       │             │
+│  │                 ◄──┬─── telemetry ◄───┐ │             │
+│  └─▲─┬─────────────┘  │     ▲            │ │             │
+│    │ │                │     │            │ │             │
+│    │ │                │     │            │ │             │
+│    │ │  ┌─────────────▼───┐ │    ┌───────┴─▼───────┐     │
+│    │ │  │ az plugin       │ │    │ thin-edge JSON  │     │
```

Comment:
Representing `thin-edge JSON` as a **box** here might be a bit confusing as it might appear like an independent component/plugin where as it's just one data format used for the communication. So, in my head, it's just one of labels on these connecting **arrows**, not a box.

Or were you trying to represent a real app component that produces data in thin-edge JSON format?

Reply:
`thin-edge JSON` is represented as a box *because* the intent is to provide support for thin-edge json as an independent component/plugin.

* The thin-edge JSON box is a component that establishes the connection between the internal rust components and the external MQTT components. 
* On the north bound arrows, flow internal messages (i.e rust in-memory struct), for `Measurement`, `Alarm`, `Operation`.
* On the south bound arrows, flow external messages (a topic + a JSON payload).
* The thin-edge JSON box is an active component that consumes messages on one side and publishes translated messages the other side.
* The internal components (say a c8y mapper or a sm plugin) have then no dependencies to thin-edge JSON nor MQTT. They process `Measurement` or `Operation`. 
* The thin-edge JSON box can be replaced by another box (say XML over XMPP) without any code impact on the internal components.